lib/src/source_visitor.dart

```diff
@@ -2755,25 +2755,71 @@ class SourceVisitor extends ThrowingAstVisitor {
 
   @override
   void visitSwitchExpressionCase(SwitchExpressionCase node) {
+    // If the pattern is a series of `||` patterns, then flatten them out and
+    // format them like empty cases with fallthrough in a switch statement
+    // instead of like a single indented binary pattern. Prefer:
+    //
+    //   e = switch (obj) {
+    //     constant1 ||
+    //     constant2 ||
+    //     constant3 =>
+    //       body
+    //   };
+    //
+    // Instead of:
+    //
+    //   e = switch (obj) {
+    //     constant1 ||
+    //        constant2 ||
+    //        constant3 =>
+    //       body
+    //   };
+    var orBranches = <DartPattern>[];
+    var orTokens = <Token>[];
+
+    void flattenOr(DartPattern e) {
+      if (e is! LogicalOrPattern) {
+        orBranches.add(e);
+      } else {
+        flattenOr(e.leftOperand);
+        orTokens.add(e.operator);
+        flattenOr(e.rightOperand);
+      }
+    }
+
+    flattenOr(node.guardedPattern.pattern);
+
     // Wrap the rule for splitting after "=>" around the pattern so that a
     // split in the pattern forces the expression to move to the next line too.
     builder.startLazyRule();
 
-    // Wrap the expression's nesting around the pattern too so that a split in
+    // Write the "||" operands up to the last one.
+    for (var i = 0; i < orBranches.length - 1; i++) {
+      visit(orBranches[i]);
+      space();
+      token(orTokens[i]);
```

Comment:
I guess we trust that `orTokens` is no shorter than `orBranches`?

Reply:
Yes, by construction `orBranches` will always have length exactly one longer than `orTokens`. Added a comment.